server/tests/api_v4/authenticated/test_user_update.py

```diff
@@ -33,3 +42,212 @@ async def test_authenticated_user_update_ok(coolorg: CoolorgRpcClients, backend:
+    (
+        "user_update_certificate",
+        "user_update_certificate_not_author_user_id",
+        "user_update_certificate_author_device_mismatch",
+        "user_update_certificate_author_update_itself",
+    ),
+)
+async def test_authenticated_user_update_invalid_certificate(
+    coolorg: CoolorgRpcClients, backend: Backend, kind: str
+) -> None:
+    now = DateTime.now()
+
+    match kind:
+        case "user_update_certificate":
+            certif = b"<dummy>"
+        case "user_update_certificate_not_author_user_id":
+            certif = UserUpdateCertificate(
+                author=coolorg.bob.device_id,
+                timestamp=now,
+                user_id=coolorg.bob.device_id.user_id,
+                new_profile=UserProfile.ADMIN,
+            ).dump_and_sign(coolorg.alice.signing_key)
+        case "user_update_certificate_author_device_mismatch":
+            certif = UserUpdateCertificate(
+                author=DeviceID("alice@dev2"),
+                timestamp=now,
+                user_id=coolorg.bob.device_id.user_id,
+                new_profile=UserProfile.ADMIN,
+            ).dump_and_sign(coolorg.alice.signing_key)
+        case "user_update_certificate_author_update_itself":
+            certif = UserUpdateCertificate(
+                author=coolorg.alice.device_id,
+                timestamp=now,
+                user_id=coolorg.alice.device_id.user_id,
+                new_profile=UserProfile.STANDARD,
+            ).dump_and_sign(coolorg.alice.signing_key)
+        case unknown:
+            assert False, unknown
+
+    rep = await coolorg.alice.user_update(user_update_certificate=certif)
+    assert rep == authenticated_cmds.v4.user_update.RepInvalidCertificate()
+
+
+async def test_authenticated_user_update_timestamp_out_of_ballpark(
+    coolorg: CoolorgRpcClients, backend: Backend
+) -> None:
+    t0 = DateTime.now().subtract(seconds=3600)
+    certif = UserUpdateCertificate(
+        author=coolorg.alice.device_id,
+        timestamp=t0,
+        user_id=coolorg.bob.device_id.user_id,
+        new_profile=UserProfile.ADMIN,
+    )
+
+    rep = await coolorg.alice.user_update(
+        user_update_certificate=certif.dump_and_sign(coolorg.alice.signing_key)
+    )
+
+    assert isinstance(rep, authenticated_cmds.v4.user_update.RepTimestampOutOfBallpark)
+    assert rep.ballpark_client_early_offset == 300.0
+    assert rep.ballpark_client_late_offset == 320.0
+    assert rep.client_timestamp == t0
+
+
+@pytest.mark.parametrize("kind", ("same_timestamp", "previous_timestamp"))
+async def test_authenticated_user_update_require_greater_timestamp(
+    coolorg: CoolorgRpcClients,
+    backend: Backend,
+    kind: str,
+) -> None:
+    now = DateTime.now()
+    match kind:
+        case "same_timestamp":
+            user_update_timestamp = now
+        case "previous_timestamp":
+            user_update_timestamp = now.subtract(seconds=1)
+        case unknown:
+            assert False, unknown
```

Comment:
You could directly put those value in the `parametrize`

Reply:
Can you provide a code example?

Reply:
```python
@pytest.mark.parametrize(
  "timestamp_offset", 
  (
    pytest.param(0, id="same_timestamp"), 
    pytest.param(42, id="previous_timestamp")
  )
```

In the function you directly use

```python
user_update_timestamp = now.subtract(seconds=timestamp_offset)
```

Reply:
Worked like a charm. Thanks!